rust/kernel/seq_file.rs

```diff
@@ -0,0 +1,213 @@
+
+    /// Called once each time the `seq_file` is opened.
+    fn start(arg: &Self) -> Option<Self::IteratorWrapper>;
+
+    /// How the item will be displayed to the reader.
+    fn display(item: &Self::Item) -> &str;
+}
+
+extern "C" fn stop_callback<T: SeqOperations>(
+    _m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) {
+    if !v.is_null() {
+        // SAFETY: `v` was created by a previous call to `next_callback` or
+        // `start_callback` and both functions return either a null pointer
+        // or pointer generated by `T::IteratorWrapper::into_pointer`.
+        drop(unsafe { T::IteratorWrapper::from_pointer(v) })
+    }
+}
+
+extern "C" fn next_callback<T: SeqOperations>(
+    _m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+    pos: *mut bindings::loff_t,
+) -> *mut c_types::c_void {
+    if v.is_null() {
+        return ptr::null_mut();
+    }
+
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `T::IteratorWrapper::into_pointer`.
+    // We already checked for he null pointer case above.
+    let mut iterator = unsafe { T::IteratorWrapper::from_pointer(v) };
+
+    // SAFETY: The caller guarantees tha `pos` is a valid pointer to an
+    // `loff_t` and expects this function to mutate the value.
+    unsafe {
+        *pos += 1;
+    }
+
+    if iterator.next().is_none() {
+        return ptr::null_mut();
+    }
+
+    match iterator.peek() {
+        Some(_next) => T::IteratorWrapper::into_pointer(iterator) as *mut _,
+        None => ptr::null_mut(),
+    }
+}
+
+extern "C" fn show_callback<T: SeqOperations>(
+    m: *mut bindings::seq_file,
+    v: *mut c_types::c_void,
+) -> c_types::c_int {
+    const FORMAT: CStr<'static> = cstr!("%.*s");
+    if v.is_null() {
+        return 0;
+    }
+    // SAFETY: `v` was created by a previous call to `next_callback` or
+    // `start_callback` and both functions return either a null pointer
+    // or pointer generated by `T::IteratorWrapper::into_pointer`. We
+    // checked for null pointers above. The iterator is forgotten below
+    // so the pointer on the C side stays valid.
+    let mut iterator = unsafe { T::IteratorWrapper::from_pointer(v) };
+    if let Some(item) = iterator.peek() {
+        let s = T::display(item);
+        // SAFETY: Calling a C function. `FORMAT` is null terminated because
+        // it comes from a `CStr`. `s` does not need to be null terminated
+        // because we are only printing the first `s.len()` bytes.
+        unsafe {
+            bindings::seq_printf(
+                m,
+                FORMAT.as_ptr() as *const c_types::c_char,
+                s.len(),
+                s.as_ptr() as *const u8 as *const c_types::c_char,
+            );
+        }
+    }
+    // Need to forget the iterator because the C side is still using a
+    // reference to it.
+    mem::forget(iterator);
```

Comment:
I wonder if we need to improve the abstraction so that it can return a reference to the underlying object. This way the caller can use it without having to `forget` or `into_pointer` it again.

Not something for this PR though.

Reply:
Should just be additional methods on `PointerWrapper`?